keras/callbacks.py

```diff
@@ -694,13 +703,16 @@ def on_epoch_end(self, epoch, logs=None):
             if epoch % self.histogram_freq == 0:
                 # TODO: implement batched calls to sess.run
                 # (current call will likely go OOM on GPU)
+
+                val_data = self.validation_data
+                tensors = self.model.inputs + \
+                    self.model.targets + \
+                    self.model.sample_weights
+
                 if self.model.uses_learning_phase:
-                    cut_v_data = len(self.model.inputs)
-                    val_data = self.validation_data[:cut_v_data] + [0]
-                    tensors = self.model.inputs + [K.learning_phase()]
-                else:
-                    val_data = self.validation_data
-                    tensors = self.model.inputs
+                    tensors += [K.learning_phase()]
+
+                assert len(val_data) == len(tensors)
```

Comment:
This wasn't required in the previous code. Previously, `val_data` was shorter than `tensors` in the else block, and it was handled nicely in `dict(zip(tensors, val_data))`. Now the lengths must match, because we'll need targets and sample weights to calculate grads.